File: OpenDDSharp/OpenDDSharp/PublicationMatchedStatus.cpp
/*********************************************************************
This file is part of OpenDDSharp.

OpenDDSharp is a .NET wrapper for OpenDDS
Copyright (C) 2018 Jose Morato

OpenDDSharp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

OpenDDSharp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with OpenDDSharp. If not, see <http://www.gnu.org/licenses/>.
**********************************************************************/
#include "PublicationMatchedStatus.h"

OpenDDSharp::DDS::PublicationMatchedStatus::PublicationMatchedStatus(::DDS::PublicationMatchedStatus status) {
	total_count = status.total_count;
	total_count_change = status.total_count_change;
	current_count = status.current_count;
	current_count_change = status.current_count_change;
	last_subscription_handle = status.last_subscription_handle;
};

System::Int32 OpenDDSharp::DDS::PublicationMatchedStatus::TotalCount::get() {
	return total_count;
};

System::Int32 OpenDDSharp::DDS::PublicationMatchedStatus::TotalCountChange::get() {
	return total_count_change;
};

System::Int32 OpenDDSharp::DDS::PublicationMatchedStatus::CurrentCount::get() {
	return current_count;
};

System::Int32 OpenDDSharp::DDS::PublicationMatchedStatus::CurrentCountChange::get() {
	return current_count_change;
};

OpenDDSharp::DDS::InstanceHandle OpenDDSharp::DDS::PublicationMatchedStatus::LastSubscriptionHandle::get() {
	return last_subscription_handle;
};

::DDS::PublicationMatchedStatus OpenDDSharp::DDS::PublicationMatchedStatus::ToNative() {
	::DDS::PublicationMatchedStatus ret;

	ret.total_count = total_count;
	ret.total_count_change = total_count_change;
	ret.current_count = current_count;
	ret.current_count_change = current_count_change;
	ret.last_subscription_handle = last_subscription_handle;

	return ret;
}

void OpenDDSharp::DDS::PublicationMatchedStatus::FromNative(::DDS::PublicationMatchedStatus native) {
	total_count = native.total_count;
	total_count_change = native.total_count_change;
	current_count = native.current_count;
	current_count_change = native.current_count_change;
	last_subscription_handle = native.last_subscription_handle;
}

File: OpenDDSharp/OpenDDSharp.Standard/DDS/PublicationMatchedStatus.cs
/*********************************************************************
This file is part of OpenDDSharp.

OpenDDSharp is a .NET wrapper for OpenDDS
Copyright (C) 2018 Jose Morato

OpenDDSharp is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

OpenDDSharp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with OpenDDSharp. If not, see <http://www.gnu.org/licenses/>.
**********************************************************************/
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace OpenDDSharp.DDS
{
    /// <summary>
    /// The PublicationMatched status indicates that either a compatible <see cref="DataReader" /> has been matched or
    /// a previously matched <see cref="DataReader" /> has ceased to be matched.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct PublicationMatchedStatus : IEquatable<PublicationMatchedStatus>
    {
        #region Properties
        /// <summary>
        /// Gets the cumulative count of data readers that have compatibly matched this <see cref="DataWriter" />.
        /// </summary>
        public int TotalCount { get; }

        /// <summary>
        /// Gets the incremental change in the total count since the last time this status was accessed.
        /// </summary>
        public int TotalCountChange { get; }

        /// <summary>
        /// Gets the current number of data readers matched to this <see cref="DataWriter" />.
        /// </summary>
        public int CurrentCount { get; }

        /// <summary>
        /// Gets the change in the current count since the last time this status was accessed.
        /// </summary>
        public int CurrentCountChange { get; }

        /// <summary>
        /// Gets the handle for the last <see cref="DataReader" /> matched.
        /// </summary>
        public InstanceHandle LastSubscriptionHandle { get; }
        #endregion

        #region IEquatable<PublicationMatchedStatus> Members
        public bool Equals(PublicationMatchedStatus other)
        {
            return TotalCount == other.TotalCount &&
                   TotalCountChange == other.TotalCountChange &&
                   CurrentCount == other.CurrentCount &&
                   CurrentCountChange == other.CurrentCountChange &&
                   LastSubscriptionHandle == other.LastSubscriptionHandle;
        }

        public override bool Equals(object obj)
        {
            return (obj is PublicationMatchedStatus other) && Equals(other);
        }

        public override int GetHashCode()
        {
            var hashCode = 543595584;
            hashCode = (hashCode * -1521134295) + TotalCount.GetHashCode();
            hashCode = (hashCode * -1521134295) + TotalCountChange.GetHashCode();
            hashCode = (hashCode * -1521134295) + CurrentCount.GetHashCode();
            hashCode = (hashCode * -1521134295) + CurrentCountChange.GetHashCode();
            hashCode = (hashCode * -1521134295) + EqualityComparer<InstanceHandle>.Default.GetHashCode(LastSubscriptionHandle);
            return hashCode;
        }
        #endregion

        #region Operators
        public static bool operator ==(PublicationMatchedStatus left, PublicationMatchedStatus right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(PublicationMatchedStatus left, PublicationMatchedStatus right)
        {
            return !(left == right);
        }
        #endregion
    }
}
